### PR TITLE
Support gradient aggregation for tf2 keras.

### PR DESCRIPTION
**Pull request by brain-good**

### CODE REVIEW

File: horovod/_keras/__init__.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 import horovod.tensorflow as hvd
 import tensorflow as tf
 from horovod.common.gradient_aggregation import LocalGradientAggregationHelper
+from horovod.common.gradient_aggregation_v2 import LocalGradientAggregationHelperV2
 from horovod.common.tf_profile import TFProfileHelper
```
Comment (aaron276h) [inline horovod/_keras/__init__.py:26]: I think we need to make this `_make_allreduce_grads_fn` look a bit more like `_DistributedOptimizer. _allreduce` and then delete `_DistributedOptimizer. _allreduce` since we aren't using it at the moment.

File: horovod/common/gradient_aggregation_v2.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        # contains not None gradients. When performing gradient
+        # aggregation we have to remove them from the list of grads
+        # prior passing them into a tf.cond().
+        self.not_none_indexes = {}
+        self.num_none_grad_updates = 0
+
+    @tf.function
+    def compute_gradients(self, grads):
+        resulting_grads = []
+        for idx, grad in enumerate(grads):
```
Comment (aaron276h) [inline horovod/common/gradient_aggregation_v2.py:49]: We should skip this for loop for cases where `self.aggregation_frequency == 1`.

File: horovod/_keras/__init__.py [hunk 3]
```diff
@@ -86,7 +95,7 @@ def get_gradients(self, loss, params):
             self.grads = super(
                 self.__class__, self).get_gradients(loss, params)
             if self._executing_eagerly:
-                return self._allreduce(self.grads)
+                return self._agg_helper.compute_gradients(tuple(self.grads))
```
Comment (aaron276h) [inline horovod/_keras/__init__.py:98]: ? check `hvd.size() > 1` before using this?

File: horovod/common/gradient_aggregation_v2.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        # Contains the mapping of indexes of grad updates that are
+        # not None to their index in shadow vars which only
+        # contains not None gradients. When performing gradient
+        # aggregation we have to remove them from the list of grads
+        # prior passing them into a tf.cond().
+        self.not_none_indexes = {}
+        self.num_none_grad_updates = 0
+
+    @tf.function
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:46]: I mean it needs it for the same reason that `compute_gradients` needs it. It's being called by the `apply_gradients` of the optimizer.

File: horovod/common/gradient_aggregation_v2.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            ]
+        return allreduced_grads
+
+    def _clear_vars(self):
+        self.counter.assign(0)
+        for idx in self.shadow_var.keys():
+            self.shadow_var[idx].assign_add(-1 * self.shadow_var[idx])
+
+    @tf.function
+    def apply_gradients(self, apply_grads_closure, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:99]: I think it was originally done to make it look like the call of the optimizer, no longer needed though

[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:99]: the graph one needs it to properly set up the control dependency on the computed gradients and have the correct gradients to apply, this one maintains it simply for the interface to be the same

File: horovod/common/gradient_aggregation_v2.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        if self.counter == 0:
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:100]: I believe the graph is only recreated if the inputs are Python constants, not things that are inside the tf.function. It seems to me that they actually are suggesting that you use Python constants if you look at their documentation around "Effective TF2": https://www.tensorflow.org/guide/effective_tf2#use_tfconfigexperimental_run_functions_eagerly_when_debugging

File: horovod/common/gradient_aggregation_v2.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            assert len(resulting_grads) == len(self.shadow_var)
+            resulting_grads = [
+                resulting_grads[idx] if idx in self.shadow_var else None
+                for idx in range(len(resulting_grads) + self.num_none_grad_updates)
+            ]
+            assert (
+                len(resulting_grads)
+                == len(self.shadow_var) + self.num_none_grad_updates
+            )
+            self._clear_vars()
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:78]: I mean `self.counter` is a variable ;)

File: horovod/common/gradient_aggregation_v2.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        return resulting_grads
+
+    def _allreduce_grads(self, grads):
+        allreduced_grads = self._allreduce_grads(grads)
+        if self.aggregation_frequency > 1:
+            gradient_divisor = (
+                self.aggregation_frequency if self.average_aggregated_gradients else 1
+            )
+            allreduced_grads = [
+                grad / gradient_divisor if grad is not None else None for grad in grads
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:89]: similar to : https://github.com/determined-ai/horovod/pull/22#discussion_r406932719

File: horovod/common/gradient_aggregation_v2.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        # is equal to `self.aggregation_frequency`. We apply gradients when `self.counter` is
+        # equal to 0.
+        self.counter = tf.Variable(initial_value=0)
+
+        self._sparse_as_dense = sparse_as_dense
+
+        # A dictionary containing the shape of each grad.
+        # This is used when gradient aggregation frequency > 1 and
+        # there are grads that have dynamically set shapes.
+        self.grad_updated_sizes_dict = grad_updated_sizes_dict
```
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py]: did something slightly diff, lmk what you think

File: horovod/_keras/__init__.py [hunk 10]
```diff
@@ -20,22 +20,28 @@
...
-            if grad is not None:
-                avg_grad = hvd.allreduce(grad,
-                                         device_dense=device_dense,
-                                         device_sparse=device_sparse,
-                                         compression=compression)
-                averaged_gradients.append(avg_grad)
-            else:
-                averaged_gradients.append(None)
-        return averaged_gradients
+        def _allreduce(gradients):
```
Comment (aaron276h) [inline horovod/_keras/__init__.py:29]: is this supposed to be here ?

File: horovod/common/gradient_aggregation_v2.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_v2.py:35]: But we don't need to add `@tf.function` since `tf.keras` will already call as a graph function with `experimental_run_tf_function` enabled.  Just ends up being a redundant call.

File: horovod/common/gradient_aggregation_v2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_v2.py:91]: Ah yes you are absolutely right.

File: horovod/common/gradient_aggregation_v2.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_v2.py:70]: When i read the code i read it as`vars` == `shadow_vars`.

File: horovod/common/gradient_aggregation_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation_v2.py:81]: I didn't see anything about best practices in that link about whether we should use python operators vs tensorflow operators. There is an extensive writeup on the topic here: https://pgaleone.eu/tensorflow/tf.function/2019/05/10/dissecting-tf-function-part-3/#lesson-1-not-all-operators-are-created-equal. 

I do agree with your point though that it seems to apply to mostly inputs, but since who knows what TF does under the covers, all things being equal it seems better to use TF operators vs python operators when possible.

File: horovod/common/gradient_aggregation_v2.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
+import tensorflow as tf
```
Comment (aaron276h) [inline horovod/common/gradient_aggregation_v2.py:1]: probably makes sense to rename this file

File: horovod/common/gradient_aggregation_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (brain-good) [reply horovod/common/gradient_aggregation_v2.py:81]: I mean in their code snippets they don't use tf operators; however, I do see your point -- happy to move it to tf operators

File: horovod/_keras/__init__.py (hunk 10 above)
Comment (brain-good) [reply horovod/_keras/__init__.py:29]: good catch
